spinalcordtoolbox/scripts/sct_dmri_display_bvecs.py

```diff
@@ -52,37 +61,82 @@ def get_parser():
     return parser
 
 
-def plot_2dscatter(fig_handle=None, subplot=None, x=None, y=None, xlabel='X', ylabel='Y'):
+def plot_2dscatter(fig_handle=None, subplot=None, x=None, y=None, xlabel='X', ylabel='Y', bvals=None, colors=None):
     ax = fig_handle.add_subplot(subplot, aspect='equal')
     for i in range(0, len(x)):
         # if b=0, do not plot
         if not(abs(x[i]) < bzero and abs(x[i]) < bzero):
-            ax.scatter(x[i], y[i])
+            ax.scatter(x[i], y[i], color=colors[bvals[i]], alpha=0.7)
+    # plt.axis('equal')
     plt.xlabel(xlabel)
     plt.ylabel(ylabel)
-    plt.xlim([-1, 1])
-    plt.ylim([-1, 1])
+    plt.xlim([-max(bvals), max(bvals)])
+    plt.ylim([-max(bvals), max(bvals)])
     plt.grid()
 
+
+def create_custom_legend(fig, shell_colors, bvals):
+    """
+    Create single custom legend for whole figure
+    :param fig: figure the legend will be created for
+    :param shell_colors: dict with b-values and colors
+    :param bvals: ndarray with all b-values
+    :return:
+    """
+
+    # count number of bvals for individual shells
+    bvals_per_shell = pd.Series(bvals[bvals > bzero]).value_counts()
+
+    # Create single custom legend for whole figure with several subplots
+    # Loop across legend elements
+    lines = [Line2D([0], [0], color=color, marker='o', markersize=8, alpha=0.7, linestyle='')
+             for color in shell_colors.values()]
+    labels = [f'b-values = {shell} (n = {bvals_per_shell[shell]})'
+              for shell in shell_colors.keys()]
+
+    # Insert legend below subplots, NB - this line has to be run after the plt.tight_layout() which is called before
+    # function call
+    legend = fig.legend(lines, labels, loc='lower left', bbox_to_anchor=(0.1, 0),
+                        bbox_transform=plt.gcf().transFigure, ncol=len(lines), fontsize=10)
```

Comment:
Currently, the legend doesn't wrap, so if there are a lot of entries, it will go off the image. :sweat_smile: 

You can see examples of this in <insert link to other review comment>

Reply:
Yeah. Thanks for this catch! I have fixed this by setting fix number of columns in the legend to 3. See, https://github.com/spinalcordtoolbox/spinalcordtoolbox/pull/3390/commits/323fa5d8f8389043525e9ebf240851fb91800ba2

Reply:
Splendid! :)

Reply:
@joshuacwnewton Hm, however I have spotted another glitch regarding to the legend 🙉  When the legend contains more than 5-6 rows, it overlaps axis in figure. It is possibly caused by fixed position of the legend relative to the axis:

https://github.com/spinalcordtoolbox/spinalcordtoolbox/blob/323fa5d8f8389043525e9ebf240851fb91800ba2/spinalcordtoolbox/scripts/sct_dmri_display_bvecs.py#L105

<details> 
<summary> see here </summary>

![bvecs](https://user-images.githubusercontent.com/39456460/118931903-becf3800-b947-11eb-812a-a54266ba67b0.png)

</details> 

I have tried to play with `tight_layout()` command, but without success...
But again, I don't assume that so many bvals will be commonly passed.

Reply:
> But again, I don't assume that so many bvals will be commonly passed.

In that case, perhaps we can leave it as-is for now, but add a comment explaining the limitation (e.g. maybe in the docstring of the legend function) to refresh our memory if we ever need to return to this in the future (e.g. if a user runs into this issue, then mentions it on the forum).

Reply:
Thanks 👍🏻 I added this limitation into docstring in https://github.com/spinalcordtoolbox/spinalcordtoolbox/pull/3390/commits/a497719eec1164e5b8de57768eb7e20d9f12b85a